### PR TITLE
Add fed IDP_ID column to IDN_FED_AUTH_SESSION_MAPPING

### LINKED ISSUE #15649: NPE while processing SLO request from federated IDP when the same IDP is configured twice.
**Describe the issue:**
NPE while processing SLO requests from federated IDP when the same IDP is configured twice., i.e. On the federated IDP side, two service providers are registered to represent the IDP with different "IdP Entity ID Alias". 2 IDPs are registered to correspond to those 2 service providers on the Identity server. The following NPE is thrown while processing the SLO request from the federated IDP.

```
[2023-03-24 07:56:44,316] [31850138-9780-438e-b60b-cc082479c6b5] ERROR {org.apache.catalina.core.ContainerBase.[Catalina].[localhost].[/].[bridgeservlet]} - Servlet.service() for servlet [bridgeservlet] in context with path [/] threw exception org.wso2.carbon.identity.application.authentication.framework.inbound.FrameworkRuntimeException: java.lang.NullPointerException
	at org.wso2.carbon.identity.application.authentication.framework.inbound.FrameworkRuntimeException.error(FrameworkRuntimeException.java:38)
	at org.wso2.carbon.identity.application.authentication.framework.inbound.IdentityServlet.getIdentityResponseFactory(IdentityServlet.java:216)
	at org.wso2.carbon.identity.application.authentication.framework.inbound.IdentityServlet.process(IdentityServlet.java:102)
	at org.wso2.carbon.identity.application.authentication.framework.inbound.IdentityServlet.service(IdentityServlet.java:44)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:764)
	at org.eclipse.equinox.http.helper.ContextPathServletAdaptor.service(ContextPathServletAdaptor.java:37)
	at org.eclipse.equinox.http.servlet.internal.ServletRegistration.service(ServletRegistration.java:61)
	at org.eclipse.equinox.http.servlet.internal.ProxyServlet.processAlias(ProxyServlet.java:128)
	at org.eclipse.equinox.http.servlet.internal.ProxyServlet.service(ProxyServlet.java:68)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:764)
	at org.wso2.carbon.tomcat.ext.servlet.DelegationServlet.service(DelegationServlet.java:68)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:227)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:53)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
	at org.owasp.csrfguard.CsrfGuardFilter.doFilter(CsrfGuardFilter.java:88)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
	at org.apache.catalina.filters.HttpHeaderSecurityFilter.doFilter(HttpHeaderSecurityFilter.java:126)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
	at org.wso2.carbon.tomcat.ext.filter.CharacterSetFilter.doFilter(CharacterSetFilter.java:65)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
	at org.apache.catalina.filters.HttpHeaderSecurityFilter.doFilter(HttpHeaderSecurityFilter.java:126)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:197)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:97)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:541)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:135)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92)
	at org.wso2.carbon.identity.context.rewrite.valve.TenantContextRewriteValve.invoke(TenantContextRewriteValve.java:86)
	at org.wso2.carbon.identity.authz.valve.AuthorizationValve.invoke(AuthorizationValve.java:110)
	at org.wso2.carbon.identity.auth.valve.AuthenticationValve.invoke(AuthenticationValve.java:111)
	at org.wso2.carbon.tomcat.ext.valves.CompositeValve.continueInvocation(CompositeValve.java:101)
	at org.wso2.carbon.tomcat.ext.valves.TomcatValveContainer.invokeValves(TomcatValveContainer.java:49)
	at org.wso2.carbon.tomcat.ext.valves.CompositeValve.invoke(CompositeValve.java:62)
	at org.wso2.carbon.tomcat.ext.valves.CarbonStuckThreadDetectionValve.invoke(CarbonStuckThreadDetectionValve.java:145)
	at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:687)
	at org.wso2.carbon.tomcat.ext.valves.CarbonContextCreatorValve.invoke(CarbonContextCreatorValve.java:59)
	at org.wso2.carbon.tomcat.ext.valves.RequestEncodingValve.invoke(RequestEncodingValve.java:49)
	at org.wso2.carbon.tomcat.ext.valves.RequestCorrelationIdValve.invoke(RequestCorrelationIdValve.java:124)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:78)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:360)
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:399)
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65)
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:890)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1743)
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
	at org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1191)
	at org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:659)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.NullPointerException
	at org.wso2.carbon.identity.application.authenticator.samlsso.logout.processor.SAMLLogoutRequestProcessor.process(SAMLLogoutRequestProcessor.java:124)
	at org.wso2.carbon.identity.application.authenticator.samlsso.logout.processor.SAMLLogoutRequestProcessor.process(SAMLLogoutRequestProcessor.java:76)
	at org.wso2.carbon.identity.application.authentication.framework.inbound.IdentityProcessCoordinator.process(IdentityProcessCoordinator.java:41)
	at org.wso2.carbon.identity.application.authentication.framework.inbound.IdentityServlet.process(IdentityServlet.java:84)
	... 52 more
```

**How to reproduce:**

- Take two 5.10.0 Identity servers, one would act as the federated IDP hence set the hostname of that Identity server as idp.com and add a host entry for it. If you are running the federated IDP also in the same server offset the port by 1.
- Apply the fixes provided in https://github.com/wso2/product-is/issues/13465, i.e. create the new columns in IDN_FED_AUTH_SESSION_MAPPING table(the issue is reproducible even without this, however just want to make sure the fix for this issue addresses the flow where the columns are present as well).
- In the federated IDP create two SAML service providers with different "IdP Entity ID Alias"(this is important because the same IDP  cannot be registered as two IDPs on the first Identity server without this feature.)
- In the service providers created configure the SLO request URL to point to the first IDP.(https://localhost:9443/identity/saml/slo?tenantDomain=carbon.super)[1]
- On the first Identity server(acting as the local IDP), create two IDPs representing the two service providers created on the federated IDP. You can download and import the metadata from the federated IDP but make sure to change the "Identity Provider Entity ID" to the  "IdP Entity ID Alias" set in the individual service providers, otherwise it would not be possible to register the same IDP as two IDPs.
- On the first Identity server, create two OIDC service providers(app1 and app2) and attach one IDP created in the previous step to one service provider.
- Now login to both service provider app1 and app2, and the logins would work successfully. Both times the user would be redirected to the federated IDP and on the federated IDP side since there is an active session, they would be able to log in automatically(SSO).
- Issue an OIDC logout request from app1, you will see the above NPE thrown and a 500 internal server error would be observed in the browser.

**Expected behavior:**
The logout should work successfully without any errors.

Analysis:
The NPE exception occurs because the sessionDetails are not returned[2] and consecutively the FederatedIdPConfigs are not set[3] causing the NPE[4]. The sessionDetails are not returned because the initially when the oidc logout request is sent the first Identity server the entry that is stored in the IDN_FED_AUTH_SESSION_MAPPING is removed[5]. There aren't two entries added against the two IDPs to the  IDN_FED_AUTH_SESSION_MAPPING table.

Note: This issue would only occur if the federated IDP has the "IdP Entity ID Alias" feature, so in multi-tenant use case scenarios and cases where the federated IDP is another Identity server this can be reproduced.

**Environment information** **:**
 - Product Version: [IS 5.10.0]

[1]https://is.docs.wso2.com/en/5.10.0/learn/handling-saml-single-logout-requests-from-federated-identity-providers/
[2]https://github.com/wso2-extensions/identity-outbound-auth-samlsso/blob/1f9f2bab3e3e5ed26756142ef921712031c523cd/components/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/logout/processor/SAMLLogoutRequestProcessor.java#L208
[3]https://github.com/wso2-extensions/identity-outbound-auth-samlsso/blob/1f9f2bab3e3e5ed26756142ef921712031c523cd/components/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/logout/processor/SAMLLogoutRequestProcessor.java#L229
[4]https://github.com/wso2-extensions/identity-outbound-auth-samlsso/blob/1f9f2bab3e3e5ed26756142ef921712031c523cd/components/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/logout/processor/SAMLLogoutRequestProcessor.java#L118
[5]https://github.com/wso2/carbon-identity-framework/blob/b4d45fe00508f910b51ef5cab72f92a3049a657b/components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultLogoutRequestHandler.java#L137

### PR DESCRIPTION
**Pull request by ThumulaPerera**
## Changes Proposed in this PR

- Fixes : https://github.com/wso2/product-is/issues/15649 
- Added the support to extract the session mapping details by using the federated IDP ID. This will be activated if the IDP_ID column is available in the IDN_FED_AUTH_SESSION_MAPPING table.
- When extracting the IDP details, the search in the IDN_FED_AUTH_SESSION_MAPPING table will take IDP_ID into consideration as well.
- IDP details will only be resolved if, Issuer value in the SAML logout request corresponds to an entry in IDN_FED_AUTH_SESSION_MAPPING table. (i.e IDP_SESSION_ID and IDP_ID will both be taken into consideration). This allows multiple IDPs to have the same IDP_SESSION_ID.
- Added new unit test cases to test the flow.
- Refactored unit tests to reuse code.

## Related PRs

Framework fix which was done to introduce the IDP ID column and storing the session details with the IDP ID[1].

## When should this PR be merged

This PR needs to be merged after the framework PR[1] is merged. The framework version needs to be updated before merging.

[1] https://github.com/wso2/carbon-identity-framework/pull/4558

### CODE REVIEW

**Comment by ThumulaPerera** [comment]
PR builder is currently failing because the framework PR[1] is not merged yet and  framework version is not updated yet.
[1] https://github.com/wso2/carbon-identity-framework/pull/4558

File: components/org.wso2.carbon.identity.application.authenticator.samlsso/src/test/java/org/wso2/carbon/identity/application/authenticator/samlsso/logout/processor/SAMLLogoutRequestProcessorTest.java [hunk 1]
```diff
@@ -88,14 +89,33 @@
...
     private static final String DB_NAME = "testSAMLSLO";
     private static final String DB_WITH_TENANT_ID_NAME = "testSAMLSLOWithTenantId";
+    private static final String DB_NAME_WITH_TENANT_ID_AND_IDP_ID_COLUMNS = "testSAMLSLOWithTenantAndIdpId";
     private static final String SAML_INDEX = "94911684-8ef8-407b-bc59-e435b6270858";
 
+    private static final String IDP_ID = "2";
+    private static final String TENANT_ID = "1";
+
+    private static final String INSERT_SQL =  "INSERT INTO IDN_FED_AUTH_SESSION_MAPPING " +
+            "(IDP_SESSION_ID, SESSION_ID, IDP_NAME,  AUTHENTICATOR_ID, PROTOCOL_TYPE) VALUES ( '" +
```
Comment (sachiniWettasinghe) [inline components/org.wso2.carbon.identity.application.authenticator.samlsso/src/test/java/org/wso2/carbon/identity/application/authenticator/samlsso/logout/processor/SAMLLogoutRequestProcessorTest.java]:
```suggestion
            "(IDP_SESSION_ID, SESSION_ID, IDP_NAME, AUTHENTICATOR_ID, PROTOCOL_TYPE) VALUES ( '" +
```

File: components/org.wso2.carbon.identity.application.authenticator.samlsso/src/test/java/org/wso2/carbon/identity/application/authenticator/samlsso/logout/processor/SAMLLogoutRequestProcessorTest.java [hunk 2]
```diff
@@ -88,14 +89,33 @@
...
+            SAML_INDEX + "' , '" + INBOUND_SESSION_INDEX + "' , '" + IDP_NAME + "' , " +
+            "'samlssoAuthenticator', 'samlsso');";
+    private static final String INSERT_SQL_WITH_TENANT_ID = "INSERT INTO IDN_FED_AUTH_SESSION_MAPPING " +
+            "(IDP_SESSION_ID, SESSION_ID, IDP_NAME,  AUTHENTICATOR_ID, PROTOCOL_TYPE, TENANT_ID) VALUES ( '" +
```
Comment (sachiniWettasinghe) [inline components/org.wso2.carbon.identity.application.authenticator.samlsso/src/test/java/org/wso2/carbon/identity/application/authenticator/samlsso/logout/processor/SAMLLogoutRequestProcessorTest.java]:
```suggestion
            "(IDP_SESSION_ID, SESSION_ID, IDP_NAME, AUTHENTICATOR_ID, PROTOCOL_TYPE, TENANT_ID) VALUES ( '" +
```

File: components/org.wso2.carbon.identity.application.authenticator.samlsso/src/test/java/org/wso2/carbon/identity/application/authenticator/samlsso/logout/processor/SAMLLogoutRequestProcessorTest.java [hunk 3]
```diff
@@ -88,14 +89,33 @@
...
+            SAML_INDEX + "' , '" + INBOUND_SESSION_INDEX + "' , '" + IDP_NAME + "' , " +
+            "'samlssoAuthenticator', 'samlsso', '" + TENANT_ID + "');";
+    private static final String INSERT_SQL_WITH_TENANT_ID_AND_IDP_ID = "INSERT INTO IDN_FED_AUTH_SESSION_MAPPING " +
+            "(IDP_SESSION_ID, SESSION_ID, IDP_NAME,  AUTHENTICATOR_ID, PROTOCOL_TYPE, TENANT_ID, IDP_ID) VALUES ( '" +
```
Comment (sachiniWettasinghe) [inline components/org.wso2.carbon.identity.application.authenticator.samlsso/src/test/java/org/wso2/carbon/identity/application/authenticator/samlsso/logout/processor/SAMLLogoutRequestProcessorTest.java]:
```suggestion
            "(IDP_SESSION_ID, SESSION_ID, IDP_NAME, AUTHENTICATOR_ID, PROTOCOL_TYPE, TENANT_ID, IDP_ID) VALUES ( '" +
```

File: components/org.wso2.carbon.identity.application.authenticator.samlsso/src/test/java/org/wso2/carbon/identity/application/authenticator/samlsso/logout/processor/SAMLLogoutRequestProcessorTest.java [hunk 4]
```diff
@@ -136,22 +156,157 @@ public void testProcess() throws Exception {
...
+        when(FrameworkUtils.isIdpIdColumnAvailableInFedAuthTable()).thenReturn(isIdpIdColumnAvailable);
+    }
+
+    private void mockIdentityTenantUtil() {
+
+        mockStatic(IdentityTenantUtil.class);
+        when(IdentityTenantUtil.getTenantId(SUPER_TENANT_DOMAIN)).thenReturn(Integer.parseInt(TENANT_ID));
+    }
+
+    private IdentityProvider getMockIdp() {
```
Comment (sachiniWettasinghe) [inline components/org.wso2.carbon.identity.application.authenticator.samlsso/src/test/java/org/wso2/carbon/identity/application/authenticator/samlsso/logout/processor/SAMLLogoutRequestProcessorTest.java:304]:
```suggestion
    private IdentityProvider getMockIdp() {
    
```

File: components/org.wso2.carbon.identity.application.authenticator.samlsso/src/test/java/org/wso2/carbon/identity/application/authenticator/samlsso/logout/processor/SAMLLogoutRequestProcessorTest.java [hunk 5]
```diff
@@ -183,50 +338,49 @@ public void testProcess() throws Exception {
...
         idp.setDefaultAuthenticatorConfig(config);
 
-        when(mockedIdPManager.getIdPByName(IDP_NAME, SUPER_TENANT_DOMAIN)).thenReturn(idp);
+        return idp;
+    }
 
-        mockStatic(SAMLSSOAuthenticatorServiceDataHolder.class);
-        when(SAMLSSOAuthenticatorServiceDataHolder.getInstance()).thenReturn(mockedAuthenticator);
-        when(mockedAuthenticator.getRealmService()).thenReturn(mockedRealmService);
+    private IdentityProvider getMockIdpWithoutSlo() {
```
Comment (sachiniWettasinghe) [inline components/org.wso2.carbon.identity.application.authenticator.samlsso/src/test/java/org/wso2/carbon/identity/application/authenticator/samlsso/logout/processor/SAMLLogoutRequestProcessorTest.java:353]:
```suggestion
    private IdentityProvider getMockIdpWithoutSlo() {
    
```

File: components/org.wso2.carbon.identity.application.authenticator.samlsso/src/test/java/org/wso2/carbon/identity/application/authenticator/samlsso/logout/processor/SAMLLogoutRequestProcessorTest.java (hunk 1 above)
Comment (ThumulaPerera) [reply components/org.wso2.carbon.identity.application.authenticator.samlsso/src/test/java/org/wso2/carbon/identity/application/authenticator/samlsso/logout/processor/SAMLLogoutRequestProcessorTest.java]: Fixed with [285e54b](https://github.com/wso2-extensions/identity-outbound-auth-samlsso/pull/161/commits/285e54b51b54c5316869f03b33d7f71dbc989d3c)

File: components/org.wso2.carbon.identity.application.authenticator.samlsso/src/test/java/org/wso2/carbon/identity/application/authenticator/samlsso/logout/processor/SAMLLogoutRequestProcessorTest.java (hunk 2 above)
Comment (ThumulaPerera) [reply components/org.wso2.carbon.identity.application.authenticator.samlsso/src/test/java/org/wso2/carbon/identity/application/authenticator/samlsso/logout/processor/SAMLLogoutRequestProcessorTest.java]: Fixed with [285e54b](https://github.com/wso2-extensions/identity-outbound-auth-samlsso/pull/161/commits/285e54b51b54c5316869f03b33d7f71dbc989d3c)

File: components/org.wso2.carbon.identity.application.authenticator.samlsso/src/test/java/org/wso2/carbon/identity/application/authenticator/samlsso/logout/processor/SAMLLogoutRequestProcessorTest.java (hunk 3 above)
Comment (ThumulaPerera) [reply components/org.wso2.carbon.identity.application.authenticator.samlsso/src/test/java/org/wso2/carbon/identity/application/authenticator/samlsso/logout/processor/SAMLLogoutRequestProcessorTest.java]: Fixed with [285e54b](https://github.com/wso2-extensions/identity-outbound-auth-samlsso/pull/161/commits/285e54b51b54c5316869f03b33d7f71dbc989d3c)

File: components/org.wso2.carbon.identity.application.authenticator.samlsso/src/test/java/org/wso2/carbon/identity/application/authenticator/samlsso/logout/processor/SAMLLogoutRequestProcessorTest.java (hunk 4 above)
Comment (ThumulaPerera) [reply components/org.wso2.carbon.identity.application.authenticator.samlsso/src/test/java/org/wso2/carbon/identity/application/authenticator/samlsso/logout/processor/SAMLLogoutRequestProcessorTest.java:304]: Fixed with [285e54b](https://github.com/wso2-extensions/identity-outbound-auth-samlsso/pull/161/commits/285e54b51b54c5316869f03b33d7f71dbc989d3c)

File: components/org.wso2.carbon.identity.application.authenticator.samlsso/src/test/java/org/wso2/carbon/identity/application/authenticator/samlsso/logout/processor/SAMLLogoutRequestProcessorTest.java (hunk 5 above)
Comment (ThumulaPerera) [reply components/org.wso2.carbon.identity.application.authenticator.samlsso/src/test/java/org/wso2/carbon/identity/application/authenticator/samlsso/logout/processor/SAMLLogoutRequestProcessorTest.java:353]: Fixed with [285e54b](https://github.com/wso2-extensions/identity-outbound-auth-samlsso/pull/161/commits/285e54b51b54c5316869f03b33d7f71dbc989d3c)

**Review by sachiniWettasinghe — DISMISSED** [review]

**Review by sachiniWettasinghe — APPROVED** [review]

**Pull request merged**